elide-core/src/main/java/com/yahoo/elide/security/executors/ActivePermissionExecutor.java

```diff
@@ -106,33 +170,37 @@ public ActivePermissionExecutor(boolean verbose, final RequestScope requestScope
      * @see com.yahoo.elide.annotation.DeletePermission
      */
     @Override
-    public <A extends Annotation> ExpressionResult checkPermission(Class<A> annotationClass,
+    public <A extends Annotation> ExpressionResult checkPermissionDeferred(EntityAction action,
                                                                    PersistentResource resource,
                                                                    ChangeSpec changeSpec) {
-        Supplier<Expression> expressionSupplier = () -> {
-            if (SharePermission.class == annotationClass) {
-                if (requestScope.getDictionary().isShareable(resource.getResourceClass())) {
-                    return expressionBuilder.buildAnyFieldExpressions(resource, ReadPermission.class, changeSpec);
-                } else {
-                    return expressionBuilder.FAIL_EXPRESSION;
-                }
-            }
-            return expressionBuilder.buildAnyFieldExpressions(resource, annotationClass, changeSpec);
-        };
+
+        if (action == EntityAction.SHARE && !requestScope.getDictionary().isShareable(resource.getResourceClass())) {
+            ForbiddenAccessException e = new ForbiddenAccessException(
+                    SharePermission.class.getSimpleName(),
+                    expressionBuilder.FAIL_EXPRESSION,
+                    Expression.EvaluationMode.ALL_CHECKS);
+            log.trace("{}", e.getLoggedMessage());
+            throw e;
```

Comment:
Could we simply return `expressionBuilder.FAIL_EXPRESSION` here instead of throwing error. Logically they might make no difference but this could change logging behavior.

Reply:
I had it done that way initially, but when we happen to have some user check on the @ReadPermission we seem to be fetching the cached expression and never seem to evaluate the FAIL_EXPRESSION that we pass. This is happening inside the checkPermissions().
https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/security/executors/ActivePermissionExecutor.java#L289

Reply:
I guess this is a side effect of mapping Annotation to EntityAction. [L143](https://github.com/yahoo/elide/blob/0f9a32893ad2394c37568a787e0da2de12df94f4/elide-core/src/main/java/com/yahoo/elide/security/executors/ActivePermissionExecutor.java#L143) get rid of annotation in the calling chain but we are still using `annotation` to check the cache in [L361](https://github.com/yahoo/elide/blob/0f9a32893ad2394c37568a787e0da2de12df94f4/elide-core/src/main/java/com/yahoo/elide/security/executors/ActivePermissionExecutor.java#L361). We should either pass the `annotation` along with `EntityAction` to preserve information loss OR change the `checkCache` to map by `EntityAction`. So that `@SharePermission` could be match to cache itself instead of `@ReadPermission`.

Reply:
Yes I think we should make changes to the caching methodology if we are going to change the model of using the PermissionExecutor. I have reverted the changes as they require better analysis with respect to caching and deferred checks. As you said this should go in a separate PR.